packages/react/src/createHooks.ts

```diff
@@ -0,0 +1,116 @@
+import { Query, Context, UpdateType, INode, LiveResult } from '@aphro/runtime-ts';
+import counter from '@strut/counter';
+import { useCallback, useMemo, useRef, useSyncExternalStore } from 'react';
+import { suspend } from 'suspend-react';
+
+const count = counter('model-infra/CreateHooks');
+
+export type ContextPromise = Promise<Context>;
+
+export type UseQueryOptions = {
+  // If you want to limit what sorts of updates will cause the live query to re-fire
+  on?: UpdateType;
+  // If you want to cache the results of the query.
+  // This'll allow components to unmount and remount without losing previously queried state.
+  // Should we just compute the key for them?
+  // We can given the declarative nature of the queries...
+  key: string;
+};
+
+type QueryReturnType<Q> = Q extends Query<infer M> ? M : any;
+
+export function createHooks(contextPromise: ContextPromise) {
+  // suspends the subtree using a hook that depends on context initialization
+  // until the context has been initialized.
+  function suspendOnContext() {
+    return suspend(() => {
+      return contextPromise;
+    }, ['context']);
+  }
+
+  function useQueryContext() {
+    const ctx = suspendOnContext();
+    return ctx;
+  }
+
+  function useQuery<Q extends Query<QueryReturnType<Q>>>(
+    queryProvider: (ctx: Context) => Q,
+    deps: any[],
+    { on, key }: UseQueryOptions,
+  ) {
+    const ctx = useQueryContext();
+
+    // allow non-stable query provider functions without rerunning dependent effects -
+    // dependencies are manually managed by the deps array parameter.
+    const providerRef = useRef(queryProvider);
+    providerRef.current = queryProvider;
+    const stableProvider = useCallback((ctx: Context) => providerRef.current(ctx), deps);
+
+    // suspend on the first result to skip 'pending' state and always work
+    // with either fresh or stale resolved data. suspended value will be cached
+    // according to user-supplied key which is required
+    // TODO: generate cache key from query
+    const liveResult = suspend(async () => {
+      count.bump('suspend.liveResult.initial');
+      const q = stableProvider(ctx);
+      const liveResult = q.live(on || UpdateType.ANY);
+      await liveResult.__currentHandle;
+      return liveResult;
+    }, [key]);
+
+    const latestResult = useSyncExternalStore(
+      cb => {
+        liveResult.subscribe(cb);
+        // FIXME: oof - need to wait til next frame to unsubscribe
+        // or the LiveResult will dispose itself.
+        return () => {
```

Comment:
`useSyncExternalStore` unsubscribes because, during suspense, the component is unmounted?

Reply:
`useSyncExternalStore` unsubscribes and then resubscribes on every store change, hence the conflict with the memory leak avoidance behavior. Each time a subscription event is detected, it evaluates referential equality of the snapshot result (second param). If the snapshot is not equal, the hook will call `unsubscribe`, then `subscribe`, then rerender the component with the new snapshot value. Before making this hack it would fail on the first change because the LiveResult will dispose of itself between `unsubscribe` and `subscribe` and throw an error when resubscribed (which is why a small deferral like `queueMicrotask` would be sufficient to avoid this).

Reply:
Gotcha. While `queueMicrotask` seems pretty fragile I'm ok with it here since leaking memory is a worse, and harder to detect, form of breakage than the alternative.